src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala

```diff
@@ -222,6 +218,30 @@ class WorkspaceService(protected val argUserToken: WithAccessToken, val rawlsDAO
 
   }
 
+  def unPublishSuccessMessage(workspaceNamespace: String, workspaceName: String): String = s" The workspace $workspaceNamespace:$workspaceName has been un-published."
+
+  def deleteWorkspace(ns: String, name: String): Future[PerRequestMessage] = {
+    rawlsDAO.getWorkspace(ns, name) flatMap { ws =>
+      val wsPublished: Boolean = isPublished(ws)
+      if (wsPublished) {
+        rawlsDAO.updateLibraryAttributes(ns, name, updatePublishAttribute(false)) flatMap { f1 =>
+          Future(removeDocument(ws.workspace, searchDAO)).flatMap { f2 =>
+            rawlsDAO.deleteWorkspace(ns, name) map { f3 =>
+              RequestComplete(f3.copy(message = Some(f3.message.getOrElse("") + unPublishSuccessMessage(ns, name))))
+            }
+          }
+        } recover {
+          case e: FireCloudExceptionWithErrorReport => RequestComplete(e.errorReport.statusCode.getOrElse(InternalServerError), e.errorReport)
+          case e => RequestComplete(InternalServerError, ErrorReport(message = e.getMessage))
+        }
+      } else {
+        rawlsDAO.deleteWorkspace(ws.workspace.namespace, ws.workspace.name) map { response =>
+          RequestComplete(response)
+        }
+      }
+    }
+  }
+
```

Comment:
this duplicates a bunch of logic with `LibraryService.setWorkspaceIsPublished`. Could you call that method instead of duplicating here? You might have to move `setWorkspaceIsPublished` into `WorkspacePublishingSupport` or `LibraryServiceSupport` to achieve the reuse.

Reply:
I started out on that road, but that method is doing much more than what I want to do and is completing the request in cases that I don't want it to. I think I can extract out the last `else` condition in that method to look like what I have.

Reply:
@davidangb - Since `setWorkspaceIsPublished` has a good chunk of preliminary logic checking that this feature doesn't need to go through, I ended up extracting what I needed from `setWorkspaceIsPublished` into another method in `WorkspacePublishingSupport`. Please have a second look.